pkg/resources/common/validation.go

```diff
@@ -58,3 +61,55 @@ func ValidateRules(rules []rbacv1.PolicyRule, isNamespaced bool, fldPath *field.
 	}
 	return returnErr
 }
+
+var annotationsFieldPath = field.NewPath("metadata").Child("annotations")
+
+// CheckCreatorPrincipalName checks that if creator-principal-name annotation is set then creatorId annotation must be set as well.
+// The value of creator-principal-name annotation should match the creator's user principal id.
+func CheckCreatorPrincipalName(userCache controllerv3.UserCache, obj metav1.Object) (*field.Error, error) {
+	annotations := obj.GetAnnotations()
+	principalName := annotations[auth.CreatorPrincipalNameAnn]
+	if principalName == "" { // Nothing to check.
+		return nil, nil
+	}
+
+	creatorID := annotations[auth.CreatorIDAnn]
+	if creatorID == "" {
+		return field.Invalid(annotationsFieldPath, auth.CreatorPrincipalNameAnn, fmt.Sprintf("annotation %s is required", auth.CreatorIDAnn)), nil
+	}
+
+	user, err := userCache.Get(creatorID)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return field.Invalid(annotationsFieldPath, auth.CreatorPrincipalNameAnn, fmt.Sprintf("creator user %s doesn't exist", creatorID)), nil
+		}
+		return nil, fmt.Errorf("error getting creator user %s: %w", creatorID, err)
+	}
+
+	for _, principal := range user.PrincipalIDs {
+		if principal == principalName {
+			return nil, nil
+		}
+	}
+
+	return field.Invalid(annotationsFieldPath, auth.CreatorPrincipalNameAnn, fmt.Sprintf("creator user %s doesn't have principal %s", creatorID, principalName)), nil
+}
+
+// CheckCreatorAnnotationsOnUpdate checks that the creatorId and creator-principal-name annotations are immutable.
+// The only allowed update is removing the annotations.
+// This function should only be called for the update operation.
+func CheckCreatorAnnotationsOnUpdate(oldObj, newObj metav1.Object) *field.Error {
+	oldAnnotations := oldObj.GetAnnotations()
+	newAnnotations := newObj.GetAnnotations()
+
+	for _, annotation := range []string{auth.CreatorIDAnn, auth.CreatorPrincipalNameAnn} {
+		if _, ok := newAnnotations[annotation]; ok {
+			// If the annotation exists on the new object it must be the same as on the old object.
+			if oldAnnotations[annotation] != newAnnotations[annotation] {
+				return field.Invalid(annotationsFieldPath, annotation, "annotation is immutable")
```

Comment:
What if both of them are changed?

Maybe we could use the `field.ErrorList`  which you can `.Aggregate()` or combine with `apierrors.NewInvalid` ?

It's usually nicer to report all validation errors in a single pass?

Reply:
This is deliberate. None of the checks for cluster and/or project are aggregated or any other object (except for PSA) for that matter. In general I agree. I'd like to have my validation errors altogether. But IMHO consistency is the key.